File: EvalVisitor.py
import math
from labeldExprVisitor import labeldExprVisitor
from labeldExprParser import labeldExprParser

class EvalVisitor(labeldExprVisitor):
    def __init__(self):
        self.memory = {}
        self.use_degrees = True   # por defecto en grados

    # --- Manejo de impresión ---
    def visitPrintExpr(self, ctx:labeldExprParser.PrintExprContext):
        value = self.visit(ctx.expr())
        print(value)
        return value

    def visitAssign(self, ctx:labeldExprParser.AssignContext):
        id = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[id] = value
        return value

    def visitId(self, ctx:labeldExprParser.IdContext):
        id = ctx.ID().getText()
        return self.memory.get(id, 0)

    def visitInt(self, ctx:labeldExprParser.IntContext):
        return int(ctx.INT().getText())

    def visitDouble(self, ctx:labeldExprParser.DoubleContext):
        return float(ctx.DOUBLE().getText())

    # --- Operaciones aritméticas ---
    def visitMulDiv(self, ctx:labeldExprParser.MulDivContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == labeldExprParser.MUL:
            return left * right
        return left / right

    def visitAddSub(self, ctx:labeldExprParser.AddSubContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == labeldExprParser.ADD:
            return left + right
        return left - right

    def visitUnaryMinus(self, ctx:labeldExprParser.UnaryMinusContext):
        return -self.visit(ctx.expr())

    def visitParens(self, ctx:labeldExprParser.ParensContext):
        return self.visit(ctx.expr())

    # --- Potencia ---
    def visitPow(self, ctx:labeldExprParser.PowContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        return math.pow(left, right)

    # --- Factorial ---
    def visitFact(self, ctx:labeldExprParser.FactContext):
        val = self.visit(ctx.expr())
        return math.factorial(int(val))

    # --- Funciones matemáticas ---
    def visitFunction(self, ctx:labeldExprParser.FunctionContext):
        func = ctx.ID().getText()
        arg = self.visit(ctx.expr()) if ctx.expr() else None

        if func == "sin": return math.sin(self._angle(arg))
        if func == "cos": return math.cos(self._angle(arg))
        if func == "tan": return math.tan(self._angle(arg))
        if func == "sqrt": return math.sqrt(arg)
        if func == "ln": return math.log(arg)
        if func == "log": return math.log10(arg)
        if func == "deg":
            self.use_degrees = True
            return 0
        if func == "rad":
            self.use_degrees = False
            return 0
        return 0

    # --- Conversión grados/radianes ---
    def _angle(self, value):
        if self.use_degrees:
            return math.radians(value)
        return value

File: EvalVisitor.java
// EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends labeldExprBaseVisitor<Double> {
    Map<String, Double> memory = new HashMap<>();
    // mode: true = degrees, false = radians (default radians)
    private boolean degreesMode = false;

    private double toRadiansIfNeeded(double x) {
        if (degreesMode) return Math.toRadians(x);
        return x;
    }

    @Override
    public Double visitAssign(labeldExprParser.AssignContext ctx) {
        String id = ctx.ID().getText();
        Double value = visit(ctx.expr());
        memory.put(id, value);
        return value;
    }

    @Override
    public Double visitPrintExpr(labeldExprParser.PrintExprContext ctx) {
        Double value = visit(ctx.expr());
        System.out.println(value);
        return 0.0;
    }

    @Override
    public Double visitInt(labeldExprParser.IntContext ctx) {
        return Double.valueOf(ctx.getText());
    }

    @Override
    public Double visitFloat(labeldExprParser.FloatContext ctx) {
        return Double.valueOf(ctx.getText());
    }

    @Override
    public Double visitId(labeldExprParser.IdContext ctx) {
        String id = ctx.ID().getText();
        if (memory.containsKey(id)) return memory.get(id);
        if (id.equalsIgnoreCase("PI")) return Math.PI;
        if (id.equalsIgnoreCase("E")) return Math.E;
        return 0.0;
    }

    @Override
    public Double visitParens(labeldExprParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    @Override
    public Double visitUnaryMinus(labeldExprParser.UnaryMinusContext ctx) {
        return -visit(ctx.expr());
    }

    @Override
    public Double visitAddSub(labeldExprParser.AddSubContext ctx) {
        Double left = visit(ctx.expr(0));
        Double right = visit(ctx.expr(1));
        if (ctx.op.getText().equals("+")) return left + right;
        return left - right;
    }

    @Override
    public Double visitMulDiv(labeldExprParser.MulDivContext ctx) {
        Double left = visit(ctx.expr(0));
        Double right = visit(ctx.expr(1));
        if (ctx.op.getText().equals("*")) return left * right;
        return left / right;
    }

    @Override
    public Double visitPowExpr(labeldExprParser.PowExprContext ctx) {
        Double base = visit(ctx.expr(0));
        Double exp = visit(ctx.expr(1));
        return Math.pow(base, exp);
    }

    @Override
    public Double visitFactorial(labeldExprParser.FactorialContext ctx) {
        Double v = visit(ctx.expr());
        return (double) factorial((int) Math.floor(v));
    }

    private long factorial(int n) {
        if (n < 0) throw new RuntimeException("factorial of negative number");
        long r = 1;
        for (int i = 2; i <= n; i++) r *= i;
        return r;
    }

    @Override
    public Double visitFunc(labeldExprParser.FuncContext ctx) {
        String name = ctx.funcCall().ID().getText();
        labeldExprParser.ArgListContext args = ctx.funcCall().argList();
        double a = 0, b = 0;
        if (args != null) {
            java.util.List<labeldExprParser.ExprContext> list = args.expr();
            if (list.size() >= 1) a = visit(list.get(0));
            if (list.size() >= 2) b = visit(list.get(1));
        }

        switch (name.toLowerCase()) {
            case "sin":
                return Math.sin(toRadiansIfNeeded(a));
            case "cos":
                return Math.cos(toRadiansIfNeeded(a));
            case "tan":
                return Math.tan(toRadiansIfNeeded(a));
            case "sqrt":
                return Math.sqrt(a);
            case "ln":
                return Math.log(a);
            case "log":
                return Math.log10(a);
            case "deg":
                degreesMode = true;
                return 0.0;
            case "rad":
                degreesMode = false;
                return 0.0;
            default:
                throw new RuntimeException("Unknown function: " + name);
        }
    }
}

